### PR TITLE
[SPARK] Support azure reads using vended credentials

### PR DESCRIPTION
**Pull request by bijanhoule**
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [ ] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

Updates the spark connector to support vended credentials for azure cloud storage (`abfs` / `abfss`).

Requires `hadoop-azure`, e.g.:
```
spark.jars.packages=org.apache.hadoop:hadoop-azure:3.3.6
```

*Note*: this currently doesn't work with hadoop-azure:3.4.0, this issue might be addressed in 3.4.1:https://issues.apache.org/jira/browse/HADOOP-19208

### CODE REVIEW

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala [hunk 1]
```diff
@@ -171,6 +171,14 @@ private class UCProxy extends TableCatalog with SupportsNamespaces {
         GcsVendedTokenProvider.ACCESS_TOKEN_KEY -> gcsCredentials.getOauthToken,
         GcsVendedTokenProvider.ACCESS_TOKEN_EXPIRATION_KEY -> temporaryCredentials.getExpirationTime.toString
       )
+    } else if (uri.getScheme == "abfs" || uri.getScheme == "abfss") {
+      val azCredentials = temporaryCredentials.getAzureUserDelegationSas
+      Map(
+        "fs.azure.account.auth.type" -> "SAS",
+        "fs.azure.account.hns.enabled" -> "true",
+        "fs.azure.sas.token.provider.type" -> "io.unitycatalog.connectors.spark.AbfsVendedTokenProvider",
+        "fs.azure.sas.fixed.token" -> azCredentials.getSasToken,
```
Comment (cloud-fan) [inline connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala:180]: shall we put these string constants in `AbfsVendedTokenProvider`?

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/AbfsVendedTokenProvider.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.apache.hadoop.fs.azurebfs.extensions.SASTokenProvider;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.security.AccessControlException;
+
+import java.io.IOException;
+
+import static java.lang.String.format;
+
+public class AbfsVendedTokenProvider implements SASTokenProvider {
```
Comment (cloud-fan) [inline connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/AbfsVendedTokenProvider.java:11]: not related to this PR, do we have document to tell uses how to set this provider in Spark configs?

**Review by cloud-fan — APPROVED** [review]

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala (hunk 1 above)
Comment (bijanhoule) [reply connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/UCSingleCatalog.scala]: Good call!  I cleaned that up, and pulled in the constants from abfs in cases where we were using already established conf values.

File: connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/AbfsVendedTokenProvider.java (hunk 2 above)
Comment (bijanhoule) [reply connectors/spark/src/main/scala/io/unitycatalog/connectors/spark/AbfsVendedTokenProvider.java:9]: Not yet -- @dennyglee and I have been working on docs for this.  I'm planning on getting that polished up once we get changes in and confirm everything is working as expected

**Pull request merged**